javascript/ql/src/Performance/ReDoS.ql

```diff
@@ -861,44 +898,215 @@ predicate isPumpable(State fork, string w) {
+    )
+  }
+
+  /**
+   * A state within a regular expression that has a pumpable state.
+   */
+  class StateInPumpableRegexp extends State {
+    pragma[noinline]
+    StateInPumpableRegexp() {
+      exists(State s | isReDoSCandidate(s, _) | getRoot(s.getRepr()) = getRoot(this.getRepr()))
+    }
+  }
+}
+
+/**
+ * Predicates for testing the presence of a rejecting suffix.
+ *
+ * These predicates are used to ensure that the all states reached from the fork
+ * by repeating `w` have a rejecting suffix.
  *
- * This predicate is used to ensure that the accepting state is not reached from the fork by repeating `w`.
- * This works under the assumption that any accepting state accepts all suffixes.
- * For example, a regexp like `/^(a+)+/` will accept any string as long the prefix is some number of `"a"`s,
- * and it is therefore not possible to construct a rejected suffix.
- * This assumption breaks on regular expression that use the anchor `$`, e.g: `/^(a+)+$/`, and such regular
- * expression are not accurately modeled by this query.
+ * For example, a regexp like `/^(a+)+/` will accept any string as long the prefix is
+ * some number of `"a"`s, and it is therefore not possible to construct a rejecting suffix.
+ *
+ * A regexp like `/(a+)+$/` or `/(a+)+b/` trivially has a rejecting suffix,
+ * as the suffix "X" will cause both the regular expressions to be rejected.
  *
  * The string `w` is repeated any number of times because it needs to be
  * infinitely repeatedable for the attack to work.
- * For a regular expression `/((ab)+)*abab/` the accepting state is not reachable from the fork
- * using epsilon transitions. But any attempt at repeating `w` will end in the accepting state.
- * This also relies on the assumption that any accepting state will accept all suffixes.
+ * For the regular expression `/((ab)+)*abab/` the accepting state is not reachable from the fork
+ * using epsilon transitions. But any attempt at repeating `w` will end in a state that accepts all suffixes.
  */
-State process(State fork, string w, int i) {
-  isPumpable(fork, w) and
-  exists(State prev |
-    i = 0 and prev = fork
+module SuffixConstruction {
+  import PrefixConstruction
+
+  /**
+   * Holds if all states reachable from `fork` by repeating `w`
+   * are rejectable by appending some suffix.
+   */
+  predicate reachesOnlyRejectableSuffixes(State fork, string w) {
+    isReDoSCandidate(fork, w) and
+    forex(State next | next = process(fork, w, w.length() - 1) | isDefinitelyRejectable(next))
+  }
+
+  /**
+   * Holds if there definitely exists a path starting from `s` that leads to the regular expression being rejected.
+   */
+  private predicate isDefinitelyRejectable(StateInPumpableRegexp s) {
+    // exists a reject edge with some char.
+    hasRejectEdge(s, _)
     or
-    prev = process(fork, w, i - 1)
+    // all edges (at least one) with some char leads to another state that is rejectable.
```

Comment:
Please update the predicate name and its documentation to clarify that this is a best-effort approximation, and does not generally provide the "definitely" guarantee the current name seems to suggest.

I cooked up a counter example, although after reading the test cases it seems you had already found similar FPs:
![graphviz (4)](https://user-images.githubusercontent.com/316427/100216810-a3d8b980-2f0a-11eb-960e-773ba7e9c2b9.png)


(All states here are accept states, or equivalently, have an epsilon-edge to the canonical accept state).

The states `A2` and `B2` are both rejectable (by `b` and `a`, respectively). They'll be recognized by the `hasRejectEdge` case.

The states `A1` and `B1` are rejectable as well, (by `ab` and `aa`, respectively). They'll be recognized by the recursive case, since reading `a` leads to (only) known definite reject states.

The state `S` is not rejectable, however, but will be recognized by the recursive case for the same reason as above: reading `a` leads only to definite rejectable states. But the two states we reach have disjoint sets of rejecting suffixes, so there is no suffix that can force rejection from `S`.

---

Finding a rejecting string in an NFA (or in a regexp) is NP-hard, so we can't realistically solve it exactly, and even a non-trivial "definitely" check is likely too hard.

Reply:
I had some of the same thoughts.  

And it's actually even harder.   
After processing the `pump` string many times, the NFA will go through some number of states when processing the `pump` string one more time (the fixpoint of processing `pump`).  
For the suffix construction to be precise, we need to find a common suffix for all the states in this fixpoint  
(otherwise the backtracking will terminate quickly). 

> although after reading the test cases it seems you had already found similar FPs:

I haven't found concrete FPs related to this.   

> Finding a rejecting string in an NFA (or in a regexp) is NP-hard, so we can't realistically solve it exactly, and even a non-trivial "definitely" check is likely too hard.

Agreed.   

I could probably find concrete suffixes for some states, by doing a search that is conceptually similar to the prefix construction. But I think we would just get many FNs instead of a few FPs.

Reply:
> I haven't found concrete FPs related to this.

Based on the comment below, I assumed this was caused by the suffix check.
```js
// GOOD - but we don't find that no suffix is rejected
var good32 = /(a+)*([^]{2,}|X)$/;
```
The double negation maybe have thrown me off. Actually I'm still struggling with that sentence 😕

Reply:
Yeah, that sentence is bad.   

And what I meant is that i haven't found any FPs in the wild.  

The problem we have with that regexp is the range `[^]{2,}`, which accepts all strings of length 2 or more.   
We don't parse ranges when constructing the NFA. So the range is absent from the NFA. 

At the fixpoint of processing the `pump` string we will be in this catch-all range, and thus there is no rejecting suffix. 

With the NFA we do have (that doesn't include the range), a rejecting suffix is trivial to find.   
So we get a FP for that test.

Reply:
(I've since found FPs related to the above in the wild).